### PR TITLE
Broker that works with Ntex runtime/server

### PR DESCRIPTION
**Pull request by vadim-kovalyov**
What's in this PR:
- Broker that works with ntex runtime/server.
- OoS 0/1 pub/sub for mqtt V3 and V5.
- Broker uses `ntex_mqtt` protocol types. 
   - This has been discussed with @arsing and we decided to use `ntex_mqtt` codec going forward and make changes to our custom server implementation to use `ntex_mqtt` as well (just codec structs). Also, meanwhile, @arsing is looking into ways how to use buffer pool with `ntex_mqtt` codec.

Since this change is getting pretty big, I'd like to get some feedback and merge it sooner than later.

The next step (in the next PR) would be to make our custom server work with `ntex_mqtt` and as a result have two "compilable" implementations to test - ntex vs custom.

There are also a low of work to be done (in the following PRs), including but not limited to:
- Offline sessions (queue outgoing messages in memory for persisted sessions)
- Multithreading support (workers > 1)
- Unit testing.
- Error handling
- Logging
- Various protocol edge cases
- Server settings/limits

### CODE REVIEW

File: broker/src/main.rs [hunk 1]
```diff
@@ -50,25 +50,22 @@ fn main() {
...
+
+    info!("Starting MQTT Broker.");
+
     let Config {
         buffer_pool: buffer_pool_config,
     } = Config::from_files().expect("could not read config");
 
     let buffer_pool = BufferPoolImpl::from_config(&buffer_pool_config);
 
-    let limits = ResourceLimits::new();
```
Comment (nyanzebra) [inline broker/src/main.rs:60]: What is the plan going forward if we are not using resource limits here?

File: mqtt-broker/src/broker.rs [hunk 2]
```diff
@@ -3,278 +3,359 @@
...
     session_table::SessionTable,
-    v3, v5, ClientId, ConnectionHandle, ResourceLimits,
+    subscription_table::{Subscription, SubscriptionTable, SubscriptionTableImpl},
+    v3::{self},
+    v5, ClientId, Error, ResourceLimits,
 };
 
+// temp hack, to be replaced with Broker struct
+#[derive(Debug, Clone)]
+pub struct BrokerImpl(Arc<Broker>);
```
Comment (nyanzebra) [inline mqtt-broker/src/broker.rs:37]: So, this is just to be used for thread safety? What is the current threading model going to look like now that we are going all in on ntex? Are we still going to have workers pinned and then share the broker with this?

File: mqtt-broker/src/broker.rs [hunk 3]
```diff
@@ -3,278 +3,359 @@
...
-            ProtoClientId::IdWithExistingSession(id) => (
-                id.into(),
-                false,
-                Some(self.limits.default_session_expiration()),
-            ),
-        };
+        let (client_id, clean_start, session_expiration) =
+            match (&connect.client_id, connect.clean_session) {
+                (client_id, true) if client_id.is_empty() => {
+                    (Uuid::new_v4().to_string().into(), true, None)
```
Comment (nyanzebra) [inline mqtt-broker/src/broker.rs:89]: How big do we really need the client id to be?

File: mqtt-broker/src/broker.rs [hunk 4]
```diff
@@ -3,278 +3,359 @@
...
-                        return_code: proto_v3::ConnectReturnCode::Refused(
-                            proto_v3::ConnectionRefusedReason::BadUserNameOrPassword,
-                        ),
-                    }))
-                    .await?;
-                conn_handle.close().await?;
-                return Ok(client_id);
+                return Ok(v3::ConnAck::fail(
+                    v3::ConnectReasonCode::BadUsernameOrPassword,
+                ));
```
Comment (nyanzebra) [inline mqtt-broker/src/broker.rs:106]: I think I see what you are trying to do with refactor, but why Ok and not Err?

File: mqtt-broker/src/broker.rs [hunk 5]
```diff
@@ -3,278 +3,359 @@
...
+            Ok(v5::ConnAck::success(client_id, false))
+        }
     }
 
     pub async fn publish_v3(
         &self,
         client_id: ClientId,
-        publish: proto_v3::Publish<P>,
+        publish: proto_v3::Publish,
     ) -> io::Result<()> {
```
Comment (nyanzebra) [inline mqtt-broker/src/broker.rs:205]: Are you going to change this later to return a Publish or similar later?

File: broker/src/main.rs [hunk 6]
```diff
@@ -50,25 +50,22 @@ fn main() {
 
 #[cfg(feature = "ntex-default")]
 #[ntex::main]
-async fn main() {
+async fn main() -> std::io::Result<()> {
+    env_logger::init();
+
+    info!("Starting MQTT Broker.");
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply broker/src/main.rs:56]: It make sense. Better logging is one of the follow up tasks.

File: server/src/ntex/mod.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+use buffer_pool::BufferPoolImpl;
+
+pub mod v3;
+
+pub mod v5;
+
+pub fn make_stream_service_factory(
+    buffer_pool: Rc<BufferPoolImpl>,
+) -> impl StreamServiceFactory<TcpStream> {
+    drop(buffer_pool);
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply server/src/ntex/mod.rs:29]: Yes, @arsing is looking into how to use our buffer pool with ntex codec. And once we have that - this code will be fixed.

File: mqtt-broker/src/broker.rs [hunk 8]
```diff
@@ -3,278 +3,359 @@
...
-        };
+        let (client_id, clean_start, session_expiration) =
+            match (&connect.client_id, connect.clean_session) {
+                (client_id, true) if client_id.is_empty() => {
+                    (Uuid::new_v4().to_string().into(), true, None)
+                }
+                // TODO: deep clone client id.
+                (client_id, true) => (client_id.clone().into(), true, None),
+                (client_id, false) => (
+                    client_id.clone().into(), // TODO: deep clone client id.
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply mqtt-broker/src/broker.rs:94]: Let's chat offline...

File: mqtt-broker/src/broker.rs (hunk 3 above)
Comment (nyanzebra) [reply mqtt-broker/src/broker.rs:89]: Maybe I am wrong, but I thought UUID can be bigger than 23 bytes as a string? So, am just curious if we can cap the length

File: mqtt-broker/src/v5/mod.rs [hunk 9]
```diff
@@ -2,109 +2,98 @@
...
+    Offline(OfflineSession),
 }
 
-impl<P> Session<P>
-where
-    P: BufferSink,
-{
+impl Session {
     /// Called when received new CONNECT packet.
     pub async fn new_connected(
```
Comment (nyanzebra) [inline mqtt-broker/src/v5/mod.rs:29]: why async if we aren't doing the conn_ack from before? do we still need the conn_ack call?

File: mqtt-broker/src/v5/mod.rs [hunk 10]
```diff
@@ -336,33 +295,32 @@ where
...
+    pub fn take_will(&mut self) -> Option<LastWill> {
         self.will.take()
     }
 
     fn can_receive(&self) -> bool {
         self.state.queued_messages.len() < self.limits.max_queued_messages()
     }
+
+    fn handle_publish(&self, publication: Publication) -> Result<()> {
+        // self.state.queued_messages.push_back(publish);
```
Comment (nyanzebra) [inline mqtt-broker/src/v5/mod.rs:311]: Is this needed?

File: mqtt-broker/src/errors.rs [hunk 11]
```diff
@@ -1,12 +1,44 @@
+use std::convert::TryFrom;
+
+use futures::channel::mpsc::TrySendError;
+use ntex_mqtt::v5::{codec::PublishAckReason, PublishAck};
 // Copyright (c) Microsoft. All rights reserved.
```
Comment (nyanzebra) [inline mqtt-broker/src/errors.rs:5]: Should this go at the top?

File: mqtt-broker/src/v5/mod.rs (hunk 10 above)
Comment (vadim-kovalyov) [reply mqtt-broker/src/v5/mod.rs]: Removed for now

**Pull request merged**